packages/patternfly-4/react-core/src/components/Checkbox/Checkbox.js

```diff
@@ -1,24 +1,62 @@
 import React from 'react';
 import styles from '@patternfly/patternfly-next/components/Check/check.css';
-import { css } from '@patternfly/react-styles';
 import PropTypes from 'prop-types';
+import { css, getModifier } from '@patternfly/react-styles';
 
 const propTypes = {
-  /** additional classes added to the Checkbox */
+  /** Additional classes added to the Checkbox. */
   className: PropTypes.string,
   /** Flag to show if the Checkbox selection is valid or invalid. */
   isValid: PropTypes.bool,
   /** Flag to show if the Checkbox is disabled. */
   isDisabled: PropTypes.bool,
   /** A callback for when the Checkbox selection changes. */
-  onChange: PropTypes.func
+  onChange: PropTypes.func,
+  /** Label text of the checkbox. */
+  label: PropTypes.oneOfType([PropTypes.string, PropTypes.node, PropTypes.func]),
+  /** Id of the checkbox. */
+  id: props => {
+    if (!props.id) {
+      return new Error('CheckBox must have an id.');
+    }
+    if (typeof props.id !== 'string') {
+      return new Error(`Id must be string. Received ${typeof props.id}.`);
+    }
+    return null;
+  },
+  /** Aria-label of the checkbox. */
+  'aria-label': props => {
+    if (typeof props.label !== 'string' && !props['aria-label']) {
+      return new Error('CheckBox must have an aria-label, when label is not string.');
+    }
+    if (props['aria-label'] && typeof props.id !== 'string') {
+      return new Error(`Aria-label must be string. Received ${typeof props.id}.`);
+    }
+    return null;
+  }
 };
 
 const defaultProps = {
   className: '',
   isValid: true,
   isDisabled: false,
-  onChange: () => undefined
+  onChange: () => undefined,
+  label: undefined,
+  id: null,
+  'aria-label': null
+};
+
+const renderLabel = (label, checkId, isDisabled) => {
+  if (typeof label === 'function') {
```

Comment:
It looks like we are only applying PF styles when the label is a string.  Can we apply style to this regardless of type? I am concerned that the labels may be unstyled, leading to inconsistencies.

Reply:
Agreed with this. If we really need to support replacing the label component used here (for now just an html label) we should follow something like `labelComponent={CustomLabel}` or `compoents={{ Label: CustomLabel }}` that would then get passed the className, htmlFor, or an other props.

If a basic function is needed the prop should be `renderLabel` and it should be passed the htmlFor and className prop to allow the user access to the styles.

Reply:
@dmiller9911 

Maybe I don't understand but I can't see any difference between this and your solution except the amount of props. 

If user wants to use a component with props instead of the label he can do something like
`<Checkbox label={<Customcomponent className="..." htmlFor="..." />}`

Reply:
I was more indicating that the function support should be removed. It does not add any value. It only complicates things. It should be `node` (covers string, rendered element, number, etc) only. That would support the follow
```jsx
<Checkbox label="<CustomLabel />" />
<Checkbox label="Label" />
const renderLabel =() => 'Some Label';
<Checkbox label={renderLabel()}
```
The only reason we would want to pass a function to label would be to pass the className and any a11y props to it
```jsx
const renderLabel = ({ className, htmlFor }) => (
  <label className={`${className} ${someCustomClass}` } htmlFor={htmlFor}>Label Text</label>
);
<Checkbox label={renderLabel}
```
Otherwise it is no differnet than the last example in the first block.

Reply:
I agree with going with `node` here and removing the function support.

Reply:
I agree with going with `node` here and removing the function support.